### ISSUE TITLE
Overfitting with V2Pipeline

### ISSUE THREAD

**Issue by Alderi-Tokori** [gcp/leela-zero#1025]
I have compared training with the old pipeline (v1) and with the v2 pipeline, and it seems the v2 pipeline makes the value head extremely more overfit than the v1 pipeline for a reduced number of games in the training data (for 9x9, I've tested with up to 50k games in the training data and noticed this happening).

To put numbers on those words, with the v1 pipeline, in 4000 steps (batch size 512, learning rate 0.005), the value head mse against training data went from 0.189 to 0.184 (and the mse against test data was reasonable, albeit obviously higher than against training data) in my test, but with the v2 pipeline, the mse against training data went all the way down to 0.03, but against test data it was up to 0.23 (so almost as bad as random).

**Comment by Alderi-Tokori** [comment]
@ywrt Pinging you in case you have some insight about this.

**Comment by Error323** [comment]
As @jkiliani referenced, I ported the code to leela-chess and am experiencing similar problems. Value head is plummeting towards 0 and accuracy is not increasing at all. It's got a severe bug somewhere, I think the data conversion between python and tensorflow is bugged, but this is speculation at this point.

**Comment by hydrogenpi** [comment]
This is bad news

**Comment by WhiteHalmos** [comment]
Not sure if @gcp is using v2 training on the training machine yet, so might not be a big deal.

**Comment by gcp** [comment]
>Not sure if @gcp is using v2 training on the training machine yet, so might not be a big deal.

I am not, precisely to avoid problems like this.

I have similar problems training the 128x20 and 192x10 nets.

**Comment by Error323** [comment]
@gcp what is your training scheme for the regular pipeline? Are you just running it continuously with some static learning rate? How much manual labor is involved for you? I find it quite time consuming at the moment.

**Comment by gcp** [comment]
>@gcp what is your training scheme for the regular pipeline? Are you just running it continuously with some static learning rate? How much manual labor is involved for you? I find it quite time consuming at the moment.

It's a simple run.py script that layers a few curl commands around the training code. It runs continuously with a static learning rate, until we notice the networks stall (I typically take that as 2/3 of the training window being from the same network with no improvement, though right now the game generation speed is so fast that I probably have to relax it).

It does a mongodump/mongorestore from zero.sjeng.org, fetches the best network, runs net_to_model to get the TF file (so I don't even save those!), dumps out the training window from mongodb (mongo_training.py), runs parse.py (which also has a few curl commands embedded to upload and schedule matches at 2^k intervals), and parse.py exists after 128k iterations.

The things not automated:
a) The server has a small, fast SSD, so I regularly need to clean out the remote DB. This used to be required like once a week, but now that we have 100k games a day, it needs to happen every 2 days, so I want to fix this.
b) Upload of training data. This is the most problematic as it would need some work scripting.
c) Upload of SGF files. This can probably be done with a simple cron script if I have some time.

>edit: have you tried training the 128x20/192x10 nets using e.g. the past 200K chunks? From scratch that is. So starting with a learning rate of 0.02 and decay with 0.1. I think this should bootstrap them nicely.

I use a slightly larger training window (500k games) as I expect to run much more steps on them, and lower the learning rate manually when I see the loss no longer drops. We start from the current best with net2net.

**Comment by kiudee** [comment]
I agree with @Error323 that increasing the learning rate slightly when starting the training of a new network can be beneficial. We have to consider that the loss landscape is constantly changing with incoming data and a higher learning rate can be useful for escaping local optima (see [A Walk with SGD](http://arxiv.org/abs/1802.08770v1) or [Visualizing the Loss Landscape of Neural Nets](http://arxiv.org/abs/1712.09913v1) for insights regarding this).

This is especially relevant when moving to a new network architecture which needs to learn to use it’s new capabilities after net2net.

**Comment by gcp** [comment]
Yes, net2net training starts at the highest learning rate. It's basically treated as "smarter random initialization". @Ttl explained this in the net2net threads.

**Comment by gcp** [comment]
I trained 128x20 with the v2 code:

2018-03-13 08:16 | b7f840b1 VS 2573eb35 | 13 : 43 (23.21%) | 56 / 400 | fail
2018-03-13 08:14 | 2f7794c1 VS 2573eb35 | 36 : 74 (32.73%) | 110 / 400 | fail
2018-03-13 08:12 | 70959052 VS 2573eb35 | 75 : 105 (41.67%) | 180 / 400 | fail

I trained 192x10 with the old code:

2018-03-13 13:13 | 4dc12a8e VS 2573eb35 | 58 : 19 (75.32%) | 77 / 400

This seems to support there's something wrong with v2.

**Comment by ywrt** [comment]
Argh! Python! Yes, you're correct, it copies the list ref instead of the list itself.

**Comment by ywrt** [comment]
Pull-request to fix this at #1034

**Comment by ywrt** [comment]
And many thanks to  Error323@ for finding this. That wouldn't have been easy :(
Now I think to think about what unittest would have caught this.

**Comment by killerducky** [comment]
@ywrt can you add some meta-data to each position about which file and line number it came from? And then measure the randomness of the final results coming out of the final dataset.shuffle().

**Comment by Error323** [comment]
>Pull-request to fix this at #1034 

Ah I already made a PR at #1032. Finding the bug was a pain indeed, mostly because I was looking at the wrong places :sweat_smile:.

>...Now I think to think about what unittest would have caught this.

I inserted a dictionary with the filename as key and a counter as value. When observing the distribution the problem became apparent quickly. Next I reduced the worker count to 1 and it showed a single chunk being used over and over.

**Comment by Error323** [comment]
Hereby I confirmed that #1032 will train properly for leela-chess now! It destroyed our previous gen in 50K steps with an elo difference of 436! Pipeline V2 is amazing, and very necessary for leela-chess, training time reduces to under 4 hrs from 16 hrs without sucking up all my cpu cores. And when our clients produce binary format it'll be even faster.

**Comment by Alderi-Tokori** [comment]
@Error323 It seems to work well over here too now, and since disk access was a bottleneck for me, I've gone from 1800pos/s to 9000pos/s during training ^^

**Comment by Error323** [comment]
Yeah I went from 1700 pos/s to 9200 pos/s and CPU usage reduced by a factor 4. Sweetness.

**Comment by roy7** [comment]
How much faster does it make training on the 19x19 Go side of things? :)

**Comment by gcp** [comment]
Given that we're not CPU limited it should make no difference at all.

**Comment by Alderi-Tokori** [comment]
I think @ywrt said it didn't change much for 19x19 because the disk access isn't the bottleneck there.

**Comment by roy7** [comment]
Ah ok. Well at least 9x9 has sped way up. :)

**Comment by Alderi-Tokori** [comment]
Closing the issue since it was fixed.
